### PR TITLE
KAFKA-15290: Handle topic-level dynamic remote storage enable configuration

### PR DESCRIPTION
**Pull request by kamalcph**
To onboard existing topics to tiered storage, bootstrap the remote-log-components when updating the dynamic `remote.storage.enable` config on the topic.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

File: core/src/test/scala/unit/kafka/server/DynamicConfigChangeTest.scala [hunk 1]
```diff
@@ -539,4 +545,52 @@ class DynamicConfigChangeUnitTest {
...
+    assertEquals(Collections.singleton(partition), leaderPartitionsArg.getValue)
+  }
+
+  @Test
+  def testDisableRemoteLogStorageOnTopic(): Unit = {
+    val topic = "test-topic"
+    val tp = new TopicPartition(topic, 0)
+    val partition: Partition = mock(classOf[Partition])
+    when(partition.isLeader).thenReturn(true)
+    when(partition.isLeader).thenReturn(false)
```
Comment (clolov) [inline core/src/test/scala/unit/kafka/server/DynamicConfigChangeTest.scala]: I was under the impression that this line will completely override the previous one, can you confirm whether the previous line does something?

File: core/src/test/scala/unit/kafka/server/DynamicConfigChangeTest.scala [hunk 2]
```diff
@@ -539,4 +545,52 @@ class DynamicConfigChangeUnitTest {
...
+    val rlm: RemoteLogManager = mock(classOf[RemoteLogManager])
+    doNothing().when(rlm).stopPartitions(ArgumentMatchers.eq(Set(tp).asJava), ArgumentMatchers.eq(true), any())
+    doNothing().when(rlm).stopPartitions(ArgumentMatchers.eq(Set(tp).asJava), ArgumentMatchers.eq(false), any())
```
Comment (clolov) [inline core/src/test/scala/unit/kafka/server/DynamicConfigChangeTest.scala]: Similar question to the above

File: core/src/test/scala/unit/kafka/server/DynamicConfigChangeTest.scala (hunk 1 above)
Comment (clolov) [reply core/src/test/scala/unit/kafka/server/DynamicConfigChangeTest.scala]: If you want the first invocation to return true and the second to return false I think the correct syntax is:
```
when(partition.isLeader).thenReturn(true).thenReturn(false)
```

**Review by clolov — COMMENTED** [review]
Thanks for this code change!

**Review by divijvaidya — COMMENTED** [review]
Do we need to implement dynamically configurable remote storage related things in a `DynamicRemoteLogManagerConfig` class similar to https://github.com/apache/kafka/pull/14243/files ? (I haven't looked into the code in detail yet so asking for your view point)

**Comment by kamalcph** [comment]
> Do we need to implement dynamically configurable remote storage related things in a `DynamicRemoteLogManagerConfig` class similar to https://github.com/apache/kafka/pull/14243/files ? (I haven't looked into the code in detail yet so asking for your view point)

I took a pass on #14243, there we are changing the broker level configs that are related to remote storage. We can extend the `DynamicRemoteLogManagerConfig` with dynamic remote log read/write quota, thread-pool size and so on.

 In this PR, we are changing the topic-level configs. `remote.storage.enable` config is applicable only at the topic level.

**Review by divijvaidya — COMMENTED** [review]
I don't think we want to implement "disablement" until KIP-950 is discussed. In early access [1], we want to support "enabling" but "disabling" is not allowed. The only option users will have is to delete the topic. The reason is we don't want to introduce a disablement change until re-enablement is discussed since it's a one way door, e.g. we may want to add a "tiered epoch" along with disablement as suggested in KIP-950.

From https://cwiki.apache.org/confluence/display/KAFKA/Kafka+Tiered+Storage+Early+Access+Release+Notes
> Disabling Tiered Storage for a Topic: Once Tiered Storage is enabled for a topic, it becomes a permanent configuration and cannot be disabled. The only way to remove the tiered storage feature from a topic is by deleting the topic itself. It is essential to be aware that deleting the topic will result in the data being removed from remote storage. Therefore, exercise caution when enabling Tiered Storage for topics, as it becomes an irreversible operation. In future versions, [KIP-950](https://cwiki.apache.org/confluence/display/KAFKA/KIP-950%3A++Tiered+Storage+Disablement) will remove this limitation and add flexibility to disable / re-enable remote storage for a topic.

**Comment by kamalcph** [comment]
@clolov @divijvaidya @showuon @satishd 

Call for review. PTAL.

**Comment by kamalcph** [comment]
Integration test for this patch was covered in [RemoteTopicCrudTest#testEnableRemoteLogOnExistingTopicTest](https://github.com/apache/kafka/blob/trunk/core/src/test/scala/integration/kafka/admin/RemoteTopicCrudTest.scala#L161).

**Review by showuon — APPROVED** [review]
LGTM!

**Review by divijvaidya — APPROVED** [review]

**Review by satishd — APPROVED** [review]
Thanks @kamalcph for the PR. LGTM

**Comment by kamalcph** [comment]
Filed KAFKA-15400 to fix the flaky RemoteIndexCacheTest

**Review by clolov — APPROVED** [review]
Thank you for the changes, they look great! The only thing which is on me to change is the message https://github.com/apache/kafka/pull/14161/files#diff-9d7d92cafcbea44968abada4825979d4eefa9d07bf49c444ce11c0c24b327127R519 `disable` -> `delete`

**Comment by satishd** [comment]
There are a few unrelated test failures. Merging to trunk.

**Comment by satishd** [comment]
Merging to 3.6 branch.

**Pull request merged**